src/main/scala/tech/cryptonomic/conseil/tezos/MetadataCaching.scala

```diff
@@ -10,25 +11,47 @@ import tech.cryptonomic.conseil.tezos.MetadataCaching._
 object MetadataCaching {
   type LastUpdated = Long
 
-  type MetadataCache[A] = Map[String, (LastUpdated, A)]
-  type AttributesCache = MetadataCache[List[Attribute]]
-  type AttributeValuesCache = MetadataCache[RadixTree[String, String]]
-  type EntitiesCache = MetadataCache[List[Entity]]
+  /** Class representing entry in cache */
+  final case class CacheEntry[A](lastUpdated: LastUpdated, value: A)
+
+  /** Class representing key in cache */
+  final case class CacheKey(key: String)
+
+  type Cache[A] = Map[CacheKey, CacheEntry[A]]
+  type AttributesCache = Cache[List[Attribute]]
+  type AttributeValuesCache = Cache[RadixTree[String, String]]
+  type EntitiesCache = Cache[List[Entity]]
 
   /** Cache initialization statuses */
-  sealed trait CachingStatus
+  sealed trait CachingStatus extends Product with Serializable
 
   case object NotStarted extends CachingStatus
 
   case object InProgress extends CachingStatus
 
   case object Finished extends CachingStatus
 
+  import cats.implicits._
+
+  /** Initializes metadata caching */
+  def empty[F[_]](implicit concurrent: Concurrent[F]): F[MetadataCaching[F]] =
+    for {
+    cachingStatus <- Ref[F].of[CachingStatus](NotStarted)
+    attributesCache <- MVar[F].empty[AttributesCache]
+    entitiesCache <- MVar[F].empty[EntitiesCache]
+    attributeValuesCache <- MVar[F].empty[AttributeValuesCache]
+  } yield new MetadataCaching[F](cachingStatus, attributesCache, entitiesCache, attributeValuesCache)
+
+  /** Returns current time in milliseconds */
+  def now: Long = System.nanoTime()
```

Comment:
There are ways to handle this too in a purely functional fashion with cats clock, yet I would leave it like this for now.
Just to take note

Reply:
ok, noted